profiler/src/main/java/com/splunk/opentelemetry/profiler/allocation/exporter/PprofAllocationEventExporter.java

```diff
@@ -0,0 +1,291 @@
+  public static final AttributeKey<String> ENCODING = stringKey("com.splunk.encoding");
+
+  @Override
+  public void flush() {
+    // Flush is called after each JFR chunk, hopefully this will keep batch sizes small enough.
+    byte[] bytes = serializePprof();
+    String encoding = allocationDataFormat.toString().toLowerCase(Locale.ROOT).replace('_', '-');
+
+    // XXX just to give an overview of exported data size
+    logger.info("Exporting {}, size {}", encoding, bytes.length);
+
+    Attributes attributes =
+        Attributes.builder()
+            .put(SOURCE_TYPE, PROFILING_SOURCE)
+            .put(DATA_TYPE, "allocation") // tells that this message is about allocation samples
+            .put(ENCODING, encoding) // data encoding
+            .build();
+
+    String body = new String(bytes, StandardCharsets.ISO_8859_1);
+    LogDataBuilder logDataBuilder =
+        LogDataBuilder.create(resource, INSTRUMENTATION_LIBRARY_INFO)
+            .setName(PROFILING_SOURCE)
+            .setEpoch(Instant.now())
+            .setBody(body)
+            .setAttributes(attributes);
+
+    logProcessor.emit(logDataBuilder.build());
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  public static class Builder {
+    private LogProcessor logProcessor;
+    private Resource resource;
+    private AllocationDataFormat allocationDataFormat;
+
+    public PprofAllocationEventExporter build() {
+      return new PprofAllocationEventExporter(this);
+    }
+
+    public Builder logProcessor(LogProcessor logsProcessor) {
+      this.logProcessor = logsProcessor;
+      return this;
+    }
+
+    public Builder resource(Resource resource) {
+      this.resource = resource;
+      return this;
+    }
+
+    public Builder allocationDataFormat(AllocationDataFormat allocationDataFormat) {
+      this.allocationDataFormat = allocationDataFormat;
+      return this;
+    }
+  }
+
+  private static class ProfileData {
+    Profile.Builder profile = Profile.newBuilder();
+    StringTable stringTable = new StringTable(profile);
+    FunctionTable functionTable = new FunctionTable(profile, stringTable);
+    LocationTable locationTable = new LocationTable(profile, functionTable);
+
+    ProfileData() {
+      profile.addSampleType(
+          ValueType.newBuilder()
+              .setType(stringTable.get("allocationSize"))
+              .setUnit(stringTable.get("bytes"))
+              .build());
+    }
+  }
+
+  // copied from
+  // https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/profiler/memory/AllocationTracker.java
+  private static class StringTable {
+    final Profile.Builder profile;
+    final Map<String, Long> table = new HashMap<>();
+    long index = 0;
+
+    StringTable(Profile.Builder profile) {
+      this.profile = profile;
+      get(""); // 0 is reserved for the empty string
+    }
+
+    long get(String str) {
```

Comment:
Likewise, will this `get()` method ever only be invoked from the same thread on a given instance of `StringTable`?

Reply:
yes